v2/tools/generator/internal/codegen/storage/conversion_graph.go

```diff
@@ -174,3 +142,83 @@ func (graph *ConversionGraph) FindNextProperty(
 
 	return astmodel.MakePropertyReference(nextType, ref.Property()), nil
 }
+
+// searchForRenamedType walks through the conversion graph looking for a match to our configured rename.
+// If no rename is configured, returns an empty type name and -1. If one is configured, either returns the found type
+// and the number of steps, or an error if not found.
+// We only look for type renames if we're starting from a storage package - no renames are needed when converting
+// from an api package to a storage package because the storage versions are always synthesized with an exact match
+// on type names.
+func (graph *ConversionGraph) searchForRenamedType(
+	name astmodel.TypeName,
+	definitions astmodel.TypeDefinitionSet) (astmodel.TypeName, int, error) {
+
+	// No configuration, or we're not looking at a storage package
+	if graph.configuration == nil || !astmodel.IsStoragePackageReference(name.PackageReference) {
+		return astmodel.EmptyTypeName, -1, nil
+	}
+
+	rename, err := graph.configuration.LookupNameInNextVersion(name)
+	if config.IsNotConfiguredError(err) {
+		// No configured rename, nothing to do
+		return astmodel.EmptyTypeName, -1, nil
+	}
+
+	// If we have any error other than a NotConfiguredError, something went wrong, and we must abort
+	if err != nil {
+		return astmodel.EmptyTypeName, -1, errors.Wrapf(
+			err,
+			"finding next type after %s",
+			name)
+	}
+
+	newType := name.WithName(rename)
+	result, steps := graph.searchForMatchingTypeImpl(newType, definitions, 1)
+
+	// Validity check on the type-rename to verify that it specifies a type that exists.
+	// If we didn't find the type, the configured rename is invalid
+	if result.IsEmpty() {
+		group, version, ok := name.PackageReference.GroupVersion()
+		if !ok {
+			panic(fmt.Sprintf("never expected external reference %s", name.PackageReference))
+		}
+
+		return astmodel.EmptyTypeName, -1, errors.Errorf(
+			"rename of %s/%s/%s invalid because no type with name %s was found in any later version",
+			group,
+			version,
+			name.Name(),
+			rename)
+	}
+
+	return result, steps, nil
+}
+
+// searchForMatchingType walks through the conversion graph looking for the next type with the same name as the one
+// provided. Returns the found type and the number of steps, if found; otherwise returns an empty type name and -1.
+func (graph *ConversionGraph) searchForMatchingType(
+	typeName astmodel.TypeName,
+	definitions astmodel.TypeDefinitionSet) (astmodel.TypeName, int) {
+	return graph.searchForMatchingTypeImpl(typeName, definitions, 1)
+}
+
+// searchForMatchingTypeImpl is the recursive implementation of searchForMatchingType.
+func (graph *ConversionGraph) searchForMatchingTypeImpl(
+	typeName astmodel.TypeName,
+	definitions astmodel.TypeDefinitionSet,
+	steps int) (astmodel.TypeName, int) {
```

Comment:
really minor: `steps` sorta sounds like it's an instruction on how many steps to use, rather than a "current state". Wondering if it might be clearer as `currentSteps` or something. Obviously this is super minor though.

Reply:
Renamed to `stepsSoFar`